lib/abci/handlers/checkTxHandlerFactory.js

```diff
@@ -2,42 +2,95 @@ const {
   abci: {
     ResponseCheckTx,
   },
+  common: {
+    KVPair,
+  },
 } = require('abci/types');
 
 const InvalidStateTransitionError = require('@dashevo/dpp/lib/stateTransition/errors/InvalidStateTransitionError');
 
+const stateTransitionTypes = require('@dashevo/dpp/lib/stateTransition//stateTransitionTypes');
+
+const AbciError = require('../errors/AbciError');
+
 const InvalidArgumentAbciError = require('../errors/InvalidArgumentAbciError');
 
 /**
  * @param {DashPlatformProtocol} dpp
- * @return {checkTxHandler}
+ * @param {RateLimiter} rateLimiter
+ * @param {boolean} rateLimiterActive
+ *
+ * @returns {checkTxHandler}
  */
-function checkTxHandlerFactory(dpp) {
+function checkTxHandlerFactory(dpp, rateLimiter, rateLimiterActive) {
   /**
    * CheckTx ABCI handler
    *
    * @typedef checkTxHandler
    *
    * @param {abci.RequestCheckTx} request
-   * @return {Promise<abci.ResponseCheckTx>}
+   * @param {BlockchainState} blockchainState
+   * @returns {Promise<abci.ResponseCheckTx>}
    */
-  async function checkTxHandler({ tx: stateTransitionByteArray }) {
+  async function checkTxHandler({ tx: stateTransitionByteArray }, blockchainState) {
     if (!stateTransitionByteArray) {
       throw new InvalidArgumentAbciError('State Transition is not specified');
     }
 
     const stateTransitionSerialized = Buffer.from(stateTransitionByteArray);
 
+    let stateTransition;
     try {
-      await dpp.stateTransition.createFromSerialized(stateTransitionSerialized);
+      stateTransition = await dpp
+        .stateTransition
+        .createFromSerialized(stateTransitionSerialized);
     } catch (e) {
       if (e instanceof InvalidStateTransitionError) {
         throw new InvalidArgumentAbciError('State Transition is invalid', { errors: e.getErrors() });
       }
-
       throw e;
     }
 
+    // apply rate limiter for document state transitions
+    if (rateLimiterActive && stateTransition.getType() === stateTransitionTypes.DOCUMENTS) {
+      const { userId } = stateTransition.documents[0];
+
+      const isBannedUser = await rateLimiter.isBannedUser(
```

Comment:
We should check it for `deliverTx` as well. ST can come with a block, so `checkTx` won't be called.